### ISSUE TITLE
High water mark option not working for pub sub

### ISSUE THREAD

**Issue by sherlockedjd** [zeromq/jeromq#588]
hi all can someone please explain with an example how to make high water mark option work for pub type sockets. I have tried but the publisher seems to publish all the messages even when HWM value is reached.can someone provide a complete example where messages are dropped when HWM is reached on pub socket. Thanks in advance. also note i have been setting this option before binding to my pub socket.

**Comment by fredoboulo** [comment]
Hi, did you have a look on the test PubSubHwmTest ?

It seems that PUB sockets seems to be not lossy, with option 

```ZMQ.setSocketOption(pub, ZMQ.ZMQ_XPUB_NODROP, true);
```
when sending messages without waiting, like:
```
ZMQ.send(pub, msg, ZMQ.ZMQ_DONTWAIT);
```
That would allow you to detect when HWM is reached.

If you could describe more precisely the usage you want of this, we may be of better help.

**Comment by sherlockedjd** [comment]
Thanks fred for your response i shall definitely try this out. Basically i just want to explore the possibility of detecting when messages are being dropped on the publisher side because of high water mark being reached i.e. when subscriber is slow it starts dropping messages and consequently on the other hand the publisher side queue also gets full and eventually publisher starts to drop messages as mentioned in zmq docs for pub type sockets.

**Comment by sherlockedjd** [comment]
I tried sending messages with dontwait flag but it appears not be working i always get the eagian flag set to 35 after sending any number of messages.also the hwm seems not to be working on pub side

heres my pub side code if any one wants to look at this and help me out

public static void main(String[] args) throws InterruptedException {
		 //  Prepare our context and publisher
        ZMQ.Context context = ZMQ.context(1);
        ZMQ.Socket publisher = context.socket(ZMQ.PUB);
        boolean receipt;
        int count=0;
        int error=0;
        publisher.setSndHWM(50);
       publisher.bind("tcp://localhost:5559");
        String message="this is message from server1";
        Scanner sc=new Scanner(System.in);
        String input="Yes";
        Socket socket =new Socket(); 
        System.out.println(ZMQ.getVersionString());
        long starttime=System.currentTimeMillis();
        for(int i=0;i<50;i++)
        //int i=0;
        //while(true)
        {
        	message="this is message from server1";
        	message+="#"+i;
        	
        //receipt=publisher.send(message);
        receipt=publisher.send(message,ZMQ.DONTWAIT);
        if(receipt)
        {
        	count++;
        }
        if(publisher.errno()==ZError.EAGAIN)
        {
        	System.out.println(publisher.errno());
        	error++;
        	
        }
        
        if(i==0||i==5000||i==10000||i==49000)
        {
         long timeinmillisecs=System.currentTimeMillis();
       	 long millis = timeinmillisecs % 1000;
       	 long second = (timeinmillisecs / 1000) % 60;
       	 long minute = (timeinmillisecs / (1000 * 60)) % 60;
       	 long hour = (timeinmillisecs/ (1000 * 60 * 60)) % 24+5;
       	 hour=hour%12;

       	    String time = String.format("%02d:%02d:%02d.%d", hour, minute, second, millis);
        	System.out.println("in sleep baby at time"+time);
           Thread.sleep(4000);
        input=sc.next();
        if(input=="Stop")
        {
        	break;
        }
      }
     // i++;
        System.out.println(message);
        }
        long stoptime=System.currentTimeMillis();
        long timetaken=stoptime-starttime;
        System.out.println("time elapsed in sending "+count+" msgs: "+timetaken+"failed"+error);
        //publisher.close ();
        //context.term ();
	}

**Comment by fredoboulo** [comment]
You did not use Socket.setXpubNoDrop, in order to disable lossy behaviour.

As a general rule, errno should be checked only in case of error. That's the current behaviour that could certainly be improved by resetting the value before any API call, but checking it only in case of not being able to send message would help your code.

**Comment by daveyarwood** [comment]
@sherlockedjd Does the previous post help?

**Comment by petaryulianov** [comment]
Yes, When this flag is set to the socket `socket.setXpubNoDrop(true);`
The HWM feature works. But What if the user wants to drop the message if the HWM is reached?
The following got from the send functions tells me that the HWM is not respected when `socket.setXpubNoDrop(false)`
`        if (lossy || dist.checkHwm()) {
            //  Send the message to all the pipes that were marked as matching
            //  in the previous step.
            if (dist.sendToMatching(msg)) {
                //  If we are at the end of multi-part message we can mark all the pipes
                //  as non-matching.
                if (!msgMore) {
                    dist.unmatch();
                }
                more = msgMore;
                return true; //  Yay, sent successfully
            }
        }
`
